mgr/cephadm: only deploy 3 mons by default when there are 4 hosts

adk3798 opened the pull request:
Fixes: https://tracker.ceph.com/issues/47234

Signed-off-by: Adam King <adking@redhat.com>

jschmid1 commented on src/pybind/mgr/cephadm/module.py:
```diff
@@ -2176,6 +2176,12 @@ def _apply_service(self, spec: ServiceSpec) -> bool:
         daemon_type = spec.service_type
         service_name = spec.service_name()
+        # if 4 hosts, only deploy 3 mons by default rather than 4
+        if service_name == 'mon' and spec.placement.default:
```
What about making this a bit more generic? 

There are multiple cases that need to be taken care of.

* one node -> one mon
* two nodes -> one mon
* three nodes -> three mons
* four nodes -> three mons
* five nodes -> five mons
* more than five -> five mons

The rule is probably `approximate to 3 and 5 but avoid even numbers`

adk3798 replied:
1 -> 1, 3 -> 3, 5 -> 5 and more than 5 -> 5 already work as is but I'll add another case for 2 -> 1.

adk3798 commented:
jenkins test make check

sebastian-philipp commented on src/python-common/ceph/deployment/service_spec.py:
```diff
@@ -149,7 +149,8 @@ def __init__(self,
                  count=None,  # type: Optional[int]
-                 host_pattern=None  # type: Optional[str]
+                 host_pattern=None,  # type: Optional[str]
+                 default=None  # type: Optional[bool]
```
can we keep this internal to cephadm? exposing this to the public is a bit strange.

adk3798 commented:
As discussed in standup, going to try a different strategy of not touching the placement spec itself and just only deploying 3 mons when there are 4 hosts along with a log message notifying that's been done

adk3798 commented:
New approach broke a bunch of scheduling tests that were tying to place 2 monitors so I changed the service type to mgr instead which I don't think should be an issue. Also added a few tests specifically for deploying an odd number of monitors when it would normally deploy an even number.

mgfritch commented on src/pybind/mgr/cephadm/schedule.py:
```diff
@@ -124,6 +124,23 @@ def place(self):
+            # is determining factor in how many mons will be placed
+            if count >= len(candidates):
+                if (len(candidates) % 2) == 0:
+                    logger.info("deploying %s monitor(s) instead of %s for better performance" % (
```
nit but the odd num of mons is for consensus / consistency (not performance) ..

mgfritch reviewed (approved):
> New approach broke a bunch of scheduling tests that were tying to place 2 monitors so I changed the service type to mgr instead which I don't think should be an issue. 

make a seperate commit for the change in test case (mon -> mgr)?

otherwise, lgtm!

jschmid1 commented on src/pybind/mgr/cephadm/schedule.py:
```diff
@@ -124,6 +124,23 @@ def place(self):
             logger.debug('Provided hosts: %s' % candidates)
             return candidates
 
+        # if asked to place even number of mons, deploy 1 less
```
I think we not _only_ need to talk about `count`.

The initial tracker defines it a bit broader with:

```
In case you have 4 hosts, cephadm will by default spawn 4 MONs.

We should probably improve this and only create 3 MONs.

```

That'd mean we need to correct the number of mons for:

* hosts
* host_placements
* count
* labels


With the current implementation this:

``` yaml
service_type: mon
placement:
  hosts:
    - host1
    - host2
    - host3
    - host4
```

would still deploy four nodes

adk3798 replied:
@jschmid1 this PR should now work with placements specs using other methods than count.

jschmid1 commented on src/pybind/mgr/cephadm/tests/test_scheduling.py:
```diff
@@ -756,3 +756,83 @@ def test_active_assignment(service_type, placement, hosts, daemons, expected):
+                                 'host1 host2 host3 host4 host5'.split(),
+                                 [],
+                                 1
+                             ),
```
The overall logic looks fine.

If you may add test cases where `daemons` already exist I'd say this is good to go.

jschmid1 reviewed (approved):
lgtm, starting tests

mgfritch commented:
green: https://pulpito.ceph.com/mgfritch-2020-11-06_14:54:25-rados:cephadm-wip-mgfritch-testing-2020-11-05-1314-distro-basic-smithi/

mgfritch commented:
@jmolmo approve?

mgfritch commented:
ping @jmolmo ?
